### ISSUE TITLE
Server state stops streamlit app execution flow and forces a rerun?

### ISSUE THREAD

**Issue by ZeroCool940711** [whitphx/streamlit-server-state#147]
Hi there, thanks for making this component, it is really useful and almost a drop-in replacement for session_state but on the server side. We are currently using it on one project we have creating a [WebUI](https://github.com/sd-webui/stable-diffusion-webui) for Stable Diffusion, we use the server-state component to keep the models loaded for multiple users on our public instances so we do not have to load the models on memory for each independent user which would make us waste a lot of resources, also for multiple sessions for a single user on their local machine, an example would be when the user needs to open a second tab to test something or if they spend sometime away and then want to run something, on those cases the default streamlit session_state would expire or if it's a new tab it will be seen as a new session so the models would be loaded again even if they are already loaded and in memory, there is where the server-state component comes into play to keep things loaded a single time between multiple session.

Now, back to the issue mentioned on the title of this issue, here is a small video of what happens when I use `session_state` from streamlit:

https://user-images.githubusercontent.com/5977640/193652417-dfe2e2b7-3c26-4f94-a7f2-9986194dce28.mp4

Then this is what happens when I use `server_state` from the server-state component:

https://user-images.githubusercontent.com/5977640/193652499-c2468768-3a3a-4d1f-ad3e-19c56131b59f.mp4

As you can see, on the first video right after the model is loaded the first iteration for the generation starts without any need for any extra input, this is the expected behavior, on the other side on the second video when using `server_state` after the models are loaded for some reason the app stops and a rerun is triggered, then in order to start the first iteration you need to click a second time on the generate button. This can be easily known because the message with the info blue text only shows when the app is first launched and there is nothing to show in there, when the models are loading this text is replaced by the loading animation and right after if the first generation is started then that text is replaced with the progress bar and progress bar info. You can see on [line 620](https://github.com/sd-webui/stable-diffusion-webui/blob/d3e22998af0a545561f0f59bf5250e77d1f53909/scripts/txt2img.py#L620) from our code that on the Text To Image (which is the one shown on the videos) the model is loaded there and then [on line 630](https://github.com/sd-webui/stable-diffusion-webui/blob/d3e22998af0a545561f0f59bf5250e77d1f53909/scripts/txt2img.py#L630) the first generation starts right after the models finish loading, line 630 is never executed on the first run because server_states trigger a rerun which causes the app to start again and never making it to there again unless the generate button is clicked a second time.

I might be missing something and not using server-state properly or this might be a bug, in any case I would appreciate if I could get some help fixing this behavior as it's quite annoying. The first time the models are loaded it's quite slow to start and can take a couple of minutes sometimes to finish loading the models so users usually just hit the generate button once and they go to do something else until the generation is done, then they come back and think the app has crashed or something bad happened because they are not seeing their expected result after a while. 
I do not really want to stop using this component as it is quite useful, it's just this specific thing that is a problem. Thanks for your time and have a good day.

**Comment by whitphx** [comment]
That is an expected behavior. `server_state` has originally been created to build something like chat-apps where an input from one user should trigger re-runs on other members' sessions so that the input is loaded on the screens on the chat members.

However, in the first place, I think [`st.experimental_singleton`](https://docs.streamlit.io/library/api-reference/performance/st.experimental_singleton) is the tool for your purpose that holds an object across all the sessions. Please try using it, and if it does not meet your needs, let me know it again.

**Comment by ZeroCool940711** [comment]
@whitphx thanks for the reply, I did check `st.expirimental_singleton` before but the problem with it is that it's a decorator and it would also require us to rewrite a lot of code to use it. For our use case it's better to have something like `server_state` that can be used anywhere and not limited to a function decorator, I also love the fact that it is a drop-in replacement for `session_state` almost all the time if not all the time so, it's easier to use for when you test things first with `session_state`, they work but then you need to move it to work across sessions like when loading a model or a global variable that has to be used, `server_state` can be implemented most of the time with one line or two, top its three lines of code so that's another really nice thing about it, so, if possible I would like to continue using it. I do understand however that I'm pushing it a bit as we're using it in a way that it was not intended to be used but I guess it's that good that it's not to use it, we can also say the same for streamlit, in our UI we are pushing it to the limits.

**Comment by whitphx** [comment]
It makes sense.

So as a workaround, what about the following code combining `st.experimental_singleton` and a mutable object (dict)?
With this example, `global_state["data"]` can be used like the server state.
```python
import streamlit as st


@st.experimental_singleton
def make_global_state():
    return {"data": None}


global_state = make_global_state()

global_state["data"] = ...  # Assign the global model
```

**Comment by ZeroCool940711** [comment]
@whitphx after some testing trying to use the code you shared before I was not able to use it, the problem is that `st.experimental_singleton` seems to conflict with `set_page_config()`, we use `set_page_config()` to make it so the page is default to the `wide` layout but if I add `st.experimental_singleton` anywhere on the app then `set_page_config()` throws an error as it's not the first streamlit command called on the app, so, for now I will continue to use server_state as it is and hope that a fix can be implemented for it so rerunning the app and interrupting the flow of it is optional and not the default behavior, if you think it's not possible to fix that then let me know, I will try to see what then I can use to replace the server-state component but I would really like to continue using it, It's pretty good, its just this issue that happens.

**Comment by whitphx** [comment]
@ZeroCool940711 
> the problem is that `st.experimental_singleton` seems to conflict with `set_page_config()`

So I think the standard solution would be fixing this problem at the Streamlit core. Could you create an issue at the Streamlit repository reporting it?

---
>  if you think it's not possible to fix that then let me know,

Currently I'm not sure if it is possible, but I may take time anyway.
If possible, do you think the following API fits your usage where a with-block is used to suppress the re-running?

```python
from streamlit_server_state import server_state, no_rerun

with no_rerun:  # Context to suppress rerunning
    server_state["some_data"] = 42
```

**Comment by whitphx** [comment]
Yes, it's not possible to provide such API from a third-party library.
Moreover, I think it conflicts the core concept of Streamlit. Re-running upon every user interaction is one of its best technical designs and it makes the app code well declarative. In this paradigm, event callbacks are more "hacks" or "escape-hatches", but not the preferred or standard way.

**Comment by ZeroCool940711** [comment]
The `no_rerun` would probably be perfect for a lot of stuff, if its not possible then I guess I will just have to make sure the code on the UI knows how to handle reruns and find another way to improve it.

**Comment by whitphx** [comment]
`no_rerun` context manager is available now with v0.15.0
